src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs

```diff
@@ -126,5 +147,124 @@ public static partial class JsonSerializer
+        }
+
+        private static void ResolveMetadataOnDictionary(MetadataPropertyName metadata, ref ReadStack state)
+        {
+            if (metadata == MetadataPropertyName.Id)
+            {
+                // Check we are not parsing into immutable.
+                if (state.Current.JsonPropertyInfo!.DictionaryConverter != null)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataCannotParsePreservedObjectIntoImmutable(state.Current.JsonPropertyInfo.DeclaredPropertyType);
+                }
+
+                if (state.Current.DictionaryHaveKeys)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataIdIsNotFirstProperty_Dictionary(ref state.Current);
+                }
+
+                state.Current.ReadMetadataValue = true;
+            }
+            else if (metadata == MetadataPropertyName.Ref)
+            {
+                if (state.Current.DictionaryHaveKeys)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataReferenceObjectCannotContainOtherProperties_Dictionary(ref state.Current);
+                }
+
+                state.Current.ShouldHandleReference = true;
+                state.Current.ReadMetadataValue = true;
+            }
+
+            state.Current.MetadataProperty = metadata;
+            state.Current.DictionaryHaveKeys = true;
+        }
+
+        private static void ResolveMetadataOnObject(ReadOnlySpan<byte> propertyName, MetadataPropertyName meta, ref ReadStack state, ref Utf8JsonReader reader, JsonSerializerOptions options)
+        {
+            if (meta == MetadataPropertyName.NoMetadata)
+            {
+                if (state.Current.IsPreservedArray)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataPreservedArrayInvalidProperty(GetValuesFromJsonPreservedReference(ref state.Current).DeclaredPropertyType);
+                }
+                // Regular property, call main logic for HandlePropertyName.
+                HandlePropertyNameDefault(propertyName, ref state, ref reader, options);
+            }
+            else if (meta == MetadataPropertyName.Id)
+            {
+                if (state.Current.ObjectHaveProperties)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataIdIsNotFirstProperty();
+                }
+
+                // TODO: Hook up JsonPropertyInfoAsString here instead.
+                // in case read of string value for this property fails.
+                JsonPropertyInfo info = JsonPropertyInfo.s_metadataProperty;
+                info.JsonPropertyName = ReadStack.s_idMetadataPropertyName;//propertyName.ToArray();
+                state.Current.JsonPropertyInfo = info;
+
+                state.Current.ReadMetadataValue = true;
+            }
+            else if (meta == MetadataPropertyName.Values)
+            {
+                Debug.Assert(propertyName.SequenceEqual(Encoding.UTF8.GetBytes("$values")));
```

Comment:
Including the `NoMetadata` branch?

Reply:
If it makes sense, sure. Otherwise, only do it where it makes sense (where the property name is known to be a constant). What will the `propertyName` be in that case? It is guaranteed to be empty or can it be any property name?

Reply:
it is guaranteed to be any property name other than $id, $ref and $values